File: algorithms/S/Shortest Common Supersequence/Shortest Common Supersequence.java
class Solution {
    public String shortestCommonSupersequence(String str1, String str2) {
        int m=str1.length();
        int n=str2.length();
        int[][] dp=new int[m+1][n+1];
        for(int i=1;i<m+1;i++){
            for(int j=1;j<n+1;j++){
                if(str1.charAt(i-1)==str2.charAt(j-1)){
                    dp[i][j]=1+dp[i-1][j-1];
                }else{
                    dp[i][j]=Math.max(dp[i][j-1],dp[i-1][j]);
                }
            }
        }
        int i=m,j=n;
        String res="";
        while(i>0 && j>0){
            if(str1.charAt(i-1)==str2.charAt(j-1)){
                res=str1.charAt(i-1)+res;
                i--;
                j--;
            }else if(dp[i-1][j]>dp[i][j-1]){
                res=str1.charAt(i-1)+res;
                i--;
            }else{
                res=str2.charAt(j-1)+res;
                j--;
            }
        }
        while(i>0){
            res=str1.charAt(i-1)+res;
            i--;
        }
        while(j>0){
            res=str2.charAt(j-1)+res;
            j--;
        }
        return res;
    }
}

File: scripts/algorithms/S/Shortest Common Supersequence/Shortest Common Supersequence.cpp
// Runtime: 22 ms (Top 70.57%) | Memory: 10.5 MB (Top 99.25%)
class Solution
{
    string LCS(string str1, string str2, int m, int n)
    {
        int t[m+1][n+1];
        string ans = "";
        for(int i = 0;i < m+1; ++i)
        {
            for(int j = 0; j< n+1; ++j)
            {
                if(i == 0 || j == 0)
                    t[i][j] = 0;
            }
        }

        for(int i = 1; i < m+1; ++i)
        {
            for(int j = 1; j < n+1; ++j)
            {
                if(str1[i-1] == str2[j-1])
                    t[i][j] = 1 + t[i-1][j-1];
                else
                    t[i][j] = max(t[i-1][j], t[i][j-1]);
            }
        }
        int i = m, j = n;
        while(i > 0 && j > 0)
        {
            if(str1[i-1] == str2[j-1])
            {
                ans.push_back(str1[i-1]);
                --i;
                --j;
            }

            else if(t[i][j-1] > t[i-1][j])
            {
                ans.push_back(str2[j-1]);
                --j;
            }

            else
            {
                ans.push_back(str1[i-1]);
                --i;
            }
        }
        while( i > 0)
        {
            ans.push_back(str1[i-1]);
            --i;
        }
        while( j > 0)
        {
            ans.push_back(str2[j-1]);
            --j;
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }

public:
    string shortestCommonSupersequence(string str1, string str2) {

        int m = str1.length();
        int n = str2.length();

        return LCS(str1, str2, m ,n);
    }
};